### PR TITLE
New Cloudware plans

### PR DESCRIPTION
**Pull request by bobwhitelock**
This document has my current understanding of the core requirements for the new version of Cloudware, my plans for how we can minimally/initially meet these requirements, and some discussion of implementation details where relevant.

Please have a look (anyone, but particularly @ste78 / @mjtko / @ColonelPanicks) and let me know
- if this matches your understanding and will solve the initial requirements we have;
- if there's anything important missing;
- if there's anything unimportant/irrelevant which we can cut for now (or indefinitely).

### CODE REVIEW

**Comment by WilliamMcCumstie** [comment]
Do we need to jump into storing everything in a `db` at this stage? There is enough work already in get the deployments playing well with the providers as it stands. The `db` is just an extra layer of work that isn't really required.

The `Deployment` object can be implemented as an `ActiveModel`, this will make transitioning to `ActiveRecord` easier at some future date

**Comment by bobwhitelock** [comment]
> Do we need to jump into storing everything in a db at this stage? There is enough work already in get the deployments playing well with the providers as it stands. The db is just an extra layer of work that isn't really required.

I don't think this should be much extra work, and in any case the database has already been set up. If anything using a database should be less work as we don't need to manually enforce constraints on the data in an ad-hoc way, we can just have database constraints, or at least AR validations. It will also be simpler to use the data in more flexible ways than trying to manipulate flat files in different ways - e.g. we can get all outputs for templating with `Output.all` rather than needing to merge the `outputs` parts of various flat files.

> The Deployment object can be implemented as an ActiveModel, this will make transitioning to ActiveRecord easier at some future date

Don't worry about this, it's already an `ActiveRecord`.

> For the deploy command do we need to specify both the provider and the template?
Could there not be /v/l/c/templates/{aws,azure}/... directories instead?

Part of the requirements mentioned recently was wanting to allow deploying arbitrary, user-specified templates, so I think it makes sense to allow users to store these however they want and just provide the path, rather than having them live in some system directory and users needing to move things to and from this in order to deploy them. This can be changed if there's a reason we want to favour just looking for templates in one location however.

File: new/docs/design/01-new-cloudware-plans.md [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+  does not already exist, and will be associated with this `Node`.
+
+  While none will be strictly required, some of these `Node` `Output`s with a
+  particular `$output_name` may be depended upon by other parts of Cloudware to
+  perform further actions related to a `Node` (and at least one `Output` with
+  any `name` will be needed for every `Node`, for Cloudware to be aware that it
+  has deployed this `Node`). For instance, a `${node_name}__id` `Output` will
+  be required to perform future actions to interact with the `Node` on the
+  cloud platform (like the `power` commands, described below), while a
+  `${node_name}__ip` action may later be used to provide a `console` command to
```
Comment (WilliamMcCumstie) [inline new/docs/design/01-new-cloudware-plans.md]: Underscores are not valid in `aws`

File: new/docs/design/01-new-cloudware-plans.md [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+   Cloudware, without special consideration needing to be made in either of
+   these stages to support the other stage (like escaping ERB tags etc.).
+
+2. The Cloudware templating system should be intentionally simplistic and just
+   needs to support the insertion of `Output` values in otherwise static
+   templates  - anything more complex than this is beyond the scope of
+   Cloudware (and is the responsibility of Underware or however else templates
+   are being created).
+
+Given these requirements, [Mustache](https://github.com/mustache/mustache) may
```
Comment (WilliamMcCumstie) [inline new/docs/design/01-new-cloudware-plans.md]: Mustache may not be a good idea in this case as `{{name}}` is to similar to `json` and will make the templates hard to write

File: new/docs/design/01-new-cloudware-plans.md [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+  One distinction between this and the `Deployment` method of inserting an
+  `Output` is that the `Node` output should be specified without the
+  `${node_name}__` prefix, i.e.  `{{ node.gpunode01.outputs.ip }}` will insert
+  the `value` of the `Output` with `name` `gpunode01__ip`, obtained when the
+  `gpunode01` `Node` was deployed.
+
+## `cloud deploy PLATFORM NAME TEMPLATE` command
+
+This command will deploy template at (relative or absolute) path `TEMPLATE`, on
+given cloud `PLATFORM`, and with given `NAME` (aside: in practise `PLATFORM`
```
Comment (WilliamMcCumstie) [inline new/docs/design/01-new-cloudware-plans.md]: Typically the user will be working within a single provider and thus the original option flag probably makes more sense. Alternatively the different provides might be split into separate tools.

**Comment by WilliamMcCumstie** [comment]
I am going to merge this through for future reference. The current iteration of cloudware is not based of these design specifications. However a few of its ideas have been incorporated

**Pull request merged**